### PR TITLE
feat(behavior_path_planner): update pull out

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

Add several pull out methods with an interface class.

- 1) shift pull out

https://user-images.githubusercontent.com/39142679/187872468-6d5057ee-e039-499b-afc7-fe0dc8052a6b.mp4

~~[old](https://user-images.githubusercontent.com/39142679/181024453-8bb13b8c-d74e-4bdf-b068-cea8215e1258.mp4)~~

- 2) geometric pull out
  - this can be used in a smaller space than shift pull out  
 
https://user-images.githubusercontent.com/39142679/181024707-3e7ca5ee-62de-4334-b9e9-ded313de1ea1.mp4

- 3) backed pull out start pose search -> backward driving -> geometric pull out (disabled by default)
  -  this can be used in a much smaller space
  
https://user-images.githubusercontent.com/39142679/181025149-8fb9fb51-9b8f-45c4-af75-27572f4fba78.mp4

Since it is not a good idea on public roads to stop in the middle of the road for geometric_pull_out, shift_path will usually be the preferred choice.
`search_priority: "efficient_path"`  (defalut)

https://user-images.githubusercontent.com/39142679/187919149-e97b9cfc-c646-46fd-a14e-300bd22954d6.mp4

## Related links

- update geometric_parallel_parking and pull_over
  - https://github.com/autowarefoundation/autoware.universe/pull/1534
- launcher
  - https://github.com/tier4/autoware_launch/pull/439
- docs
  - https://github.com/autowarefoundation/autoware.universe/pull/1587  

## Tests performed

psim and real vehicle

- 1) shift pull out

https://user-images.githubusercontent.com/39142679/184056591-f2733cf7-630b-43a3-9261-95bd8651ebb1.MOV

https://user-images.githubusercontent.com/39142679/184056607-c216c7fe-fa95-42b6-99c2-a238aaa8807e.mp4

- 2) geometric pull out

https://user-images.githubusercontent.com/39142679/184056623-0b89ab97-3aaf-4740-88ad-cce9bbc862e7.MOV

https://user-images.githubusercontent.com/39142679/184056626-894db53e-ad65-40ef-bd30-f40651c8934c.mp4

- 3) backed pull out start pose search -> backward driving -> geometric pull out (disabled by default)

https://user-images.githubusercontent.com/39142679/183327472-6102a155-66e1-4130-a3c0-8f7ab0aa5b9c.MOV

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kosuke55** [comment]
@yukkysaito @TakaHoribe 
Could you please review this PR?:pray:

File: planning/behavior_path_planner/src/scene_module/pull_out/geometric_pull_out.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  // todo: set params only once?
+  planner_.setData(planner_data_, parallel_parking_parameters_);
+  const bool found_valid_path =
+    planner_.planPullOut(start_pose, goal_pose, road_lanes, shoulder_lanes);
+  if (!found_valid_path) {
+    return {};
+  }
+
+  // collsion check with objects in shoulder lanes
+  const double collision_margin = 1.0;  // todo: make param
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/pull_out/geometric_pull_out.cpp:54]: todo

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_out/pull_out_parameters.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  double maximum_lateral_jerk;
+  double minimum_lateral_jerk;
+  double deceleration_interval;
+  // geometric pull out
+  bool enable_geometric_pull_out;
+  double geometric_pull_out_velocity;
+  double arc_path_interval;
+  double lane_departure_margin;
+  double backward_velocity;
+  double max_steer_rad;
```
Comment (kosuke55) [inline planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_out/pull_out_parameters.hpp:43]: write the reason not using vehicle_info's in README

Comment (kosuke55) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_out/pull_out_parameters.hpp:43]: pull_out_max_steer_angle

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp [hunk 3]
```diff
@@ -29,13 +29,38 @@
...
+using motion_utils::calcLongitudinalOffsetPose;
+using tier4_autoware_utils::calcOffsetPose;
 namespace behavior_path_planner
 {
 PullOutModule::PullOutModule(
   const std::string & name, rclcpp::Node & node, const PullOutParameters & parameters)
-: SceneModuleInterface{name, node}, parameters_{parameters}
+: SceneModuleInterface{name, node},
+  parameters_{parameters},
+  clock_{node.get_clock()},
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp:40]: interface already has clock_.

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp [hunk 4]
```diff
@@ -29,13 +29,38 @@
...
+      std::make_shared<GeometricPullOut>(node, parameters, getGeometricPullOutParameters()));
+  }
+  if (pull_out_planners_.empty()) {
+    RCLCPP_DEBUG(getLogger(), "Not found enabled planner");
+  }
+
+  // debug publisher
+  pull_out_start_pose_pub_ = node.create_publisher<PoseStamped>("~/pull_out/debug/backed_pose", 1);
+  full_path_pose_array_pub_ =
+    node.create_publisher<PoseArray>("~/pull_out/debug/full_path_pose_array", 1);
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp:63]: use MarkerArray publisher of behavior_path_node

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 5]
```diff
@@ -182,7 +182,7 @@ ObjectDataArray AvoidanceModule::calcAvoidanceTargetObjects(
     current_lanes, parameters_.detection_area_left_expand_dist,
     parameters_.detection_area_right_expand_dist * (-1.0));
   const auto lane_filtered_objects_index =
-    util::filterObjectsByLanelets(objects_candidate, expanded_lanelets);
+    util::filterObjectIndicesByLanelets(objects_candidate, expanded_lanelets);
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:185]: Note: just function name is changes.

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp [hunk 6]
```diff
@@ -617,150 +421,127 @@ bool PullOutModule::isInLane(
...
-    lanelet::utils::getArcCoordinates(status_.current_lanes, status_.pull_out_path.shift_point.end);
-  const bool car_is_on_goal = (arclength_shift_end.length - arclength_current.length <
-                               parameters_.pull_out_finish_judge_buffer)
-                                ? true
-                                : false;
-
-  return car_is_on_goal;
+  odometry_buffer_.push_back(planner_data_->self_odometry);
+  // Delete old data in buffer
+  while (true) {
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp:471]: rclcpp::ok()

File: planning/behavior_path_planner/src/scene_module/pull_out/shift_pull_out.cpp [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+  const auto & dynamic_objects = planner_data_->dynamic_object;
+  const auto & road_lanes = util::getCurrentLanes(planner_data_);
+  const auto & shoulder_lanes = getPullOutLanes(road_lanes, planner_data_);
+  const auto & current_pose = planner_data_->self_pose->pose;
+
+  lanelet::ConstLanelets lanes;
+  lanes.insert(lanes.end(), road_lanes.begin(), road_lanes.end());
+  lanes.insert(lanes.end(), shoulder_lanes.begin(), shoulder_lanes.end());
+
+  if (!shoulder_lanes.empty()) {
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/pull_out/shift_pull_out.cpp:52]: return boost::none when empty

File: planning/behavior_path_planner/src/scene_module/pull_out/shift_pull_out.cpp [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+      shift_path.header = planner_data_->route_handler->getRouteHeader();
+
+      return pull_out_path;
+    }
+  }
+  return boost::none;
+}
+
+std::vector<PullOutPath> ShiftPullOut::getPullOutPaths(
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/pull_out/shift_pull_out.cpp:105]: change from get to calc

File: planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_out/pull_out_parameters.hpp (hunk 2 above)
Comment (kosuke55) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_out/pull_out_parameters.hpp]: renamed in https://github.com/autowarefoundation/autoware.universe/pull/1438/commits/47013746bacac7f1a66d9076c37ee483588bef69

Comment (kosuke55) [reply planning/behavior_path_planner/include/behavior_path_planner/scene_module/pull_out/pull_out_parameters.hpp]: will update README in https://github.com/autowarefoundation/autoware.universe/pull/1587

File: planning/behavior_path_planner/src/scene_module/pull_out/geometric_pull_out.cpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/pull_out/geometric_pull_out.cpp]: fixed in https://github.com/autowarefoundation/autoware.universe/commit/47013746bacac7f1a66d9076c37ee483588bef69

File: planning/behavior_path_planner/src/scene_module/pull_out/shift_pull_out.cpp (hunk 8 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/pull_out/shift_pull_out.cpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/1438/commits/74f1feb7410fe50e81887d9563cbb9c24fe95751

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp (hunk 3 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp]: fixed in fixed in https://github.com/autowarefoundation/autoware.universe/pull/1438/commits/74f1feb7410fe50e81887d9563cbb9c24fe95751

File: planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp (hunk 6 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/scene_module/pull_out/pull_out_module.cpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/1438/commits/74f1feb7410fe50e81887d9563cbb9c24fe95751

**Comment by kosuke55** [comment]
From the advice that it is not good on public roads to stop in the middle of the road for geometric_pull_out,
I have added an option to depart on shift_path as much as possible, even when driving back.
`search_priority: "efficient_path"` 

https://user-images.githubusercontent.com/39142679/187919149-e97b9cfc-c646-46fd-a14e-300bd22954d6.mp4

**Review by TakaHoribe — APPROVED** [review]
LGTM!

**Pull request merged**